File: rt-desktop/src/main/java/xyz/redtorch/desktop/layout/base/bean/AccountFXBean.java
package xyz.redtorch.desktop.layout.base.bean;

import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.property.SimpleStringProperty;
import javafx.scene.text.Text;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import xyz.redtorch.common.util.CommonUtils;
import xyz.redtorch.pb.CoreField.AccountField;

public class AccountFXBean {

    private final static Logger logger = LoggerFactory.getLogger(AccountFXBean.class);

    private final SimpleBooleanProperty selected = new SimpleBooleanProperty(false);
    private final SimpleStringProperty accountId = new SimpleStringProperty();
    private final SimpleStringProperty code = new SimpleStringProperty();
    private final SimpleStringProperty holder = new SimpleStringProperty();
    private final SimpleStringProperty currency = new SimpleStringProperty();
    private final SimpleStringProperty balance = new SimpleStringProperty();
    private final SimpleStringProperty marginRatio = new SimpleStringProperty();
    private final SimpleStringProperty available = new SimpleStringProperty();
    private final SimpleObjectProperty<Text> todayProfit = new SimpleObjectProperty<>();
    private final SimpleObjectProperty<Text> todayProfitRatio = new SimpleObjectProperty<>();
    private final SimpleObjectProperty<Text> closeProfit = new SimpleObjectProperty<>();
    private final SimpleObjectProperty<Text> positionProfit = new SimpleObjectProperty<>();
    private final SimpleStringProperty commission = new SimpleStringProperty();
    private final SimpleStringProperty margin = new SimpleStringProperty();
    private final SimpleStringProperty preBalance = new SimpleStringProperty();
    private final SimpleStringProperty deposit = new SimpleStringProperty();
    private final SimpleStringProperty withdraw = new SimpleStringProperty();
    private final SimpleStringProperty gatewayId = new SimpleStringProperty();

    private AccountField accountField;

    public AccountFXBean(AccountField accountField, boolean selected) {
        update(accountField, selected);
    }

    public AccountField getAccountField() {
        return accountField;
    }

    public void update(AccountField newAccountField, boolean newSelected) {
        if (newAccountField == null) {
            return;
        }

        updateSelected(newSelected);

        if (accountField == null) {
            updateAccountId(newAccountField);
            updateCode(newAccountField);
            updateHolder(newAccountField);
            updateCurrency(newAccountField);
            updateGatewayId(newAccountField);
        }

        if (newAccountField != accountField) {
            updateChangeable(newAccountField);
            accountField = newAccountField;
        }
    }

    private void updateChangeable(AccountField newAccountField) {


        updateBalance(newAccountField);
        updateMarginRatio(newAccountField);
        updateAvailable(newAccountField);
        updateTodayProfit(newAccountField);
        updateTodayProfitRatio(newAccountField);
        updateCloseProfit(newAccountField);
        updatePositionProfit(newAccountField);
        updateCommission(newAccountField);
        updateMargin(newAccountField);
        updatePreBalance(newAccountField);
        updateDeposit(newAccountField);
        updateWithdraw(newAccountField);
    }

    private void updateSelected(boolean selected){
        setSelected(selected);
    }

    private void updateAccountId(AccountField newAccountField) {
        setAccountId(newAccountField.getAccountId());
    }

    private void updateCode(AccountField newAccountField) {
        setCode(newAccountField.getCode());
    }

    private void updateHolder(AccountField newAccountField) {
        setHolder(newAccountField.getHolder());
    }

    private void updateCurrency(AccountField newAccountField) {
        setCurrency(newAccountField.getCurrency().getValueDescriptor().getName());
    }

    private void updateBalance(AccountField newAccountField) {
        if (accountField == null || !CommonUtils.isEquals(accountField.getBalance(), newAccountField.getBalance())) {
            String balance = "渲染错误";
            try {
                balance = String.format("%,.2f", newAccountField.getBalance());
            } catch (Exception e) {
                logger.error("渲染错误", e);
            }
            setBalance(balance);
        }
    }

    private void updateMarginRatio(AccountField newAccountField) {
        if (accountField == null || !CommonUtils.isEquals(accountField.getBalance(), newAccountField.getBalance()) || !CommonUtils.isEquals(accountField.getMargin(), newAccountField.getMargin())) {
            String marginRatio = "渲染错误";
            try {
                marginRatio = String.format("%,.2f%%", newAccountField.getMargin() / newAccountField.getBalance() * 100);
            } catch (Exception e) {
                logger.error("渲染错误", e);
            }
            setMarginRatio(marginRatio);
        }
    }

    private void updateAvailable(AccountField newAccountField) {
        if (accountField == null || !CommonUtils.isEquals(accountField.getAvailable(), newAccountField.getAvailable())) {
            String available = "渲染错误";
            try {
                available = String.format("%,.2f", newAccountField.getAvailable());
            } catch (Exception e) {
                logger.error("渲染错误", e);
            }
            setAvailable(available);
        }
    }

    private void updateTodayProfit(AccountField newAccountField) {
        if (accountField == null || !CommonUtils.isEquals(accountField.getBalance(), newAccountField.getBalance())) {
            Text tdProfitText = new Text("渲染错误");
            try {
                double preBalance = newAccountField.getPreBalance();

                double tdProfit = newAccountField.getBalance() - preBalance - newAccountField.getDeposit() + newAccountField.getWithdraw();

                tdProfitText.setText(String.format("%,.2f", tdProfit));
                if (tdProfit > 0) {
                    tdProfitText.getStyleClass().add("trade-long-color");
                } else if (tdProfit < 0) {
                    tdProfitText.getStyleClass().add("trade-short-color");
                }
                tdProfitText.setUserData(tdProfit);

            } catch (Exception e) {
                logger.error("渲染错误", e);
            }
            setTodayProfit(tdProfitText);
        }
    }

    private void updateTodayProfitRatio(AccountField newAccountField) {
        if (accountField == null || !CommonUtils.isEquals(accountField.getBalance(), newAccountField.getBalance())) {
            Text tdProfitRatioText = new Text("渲染错误");
            try {
                double preBalance = newAccountField.getPreBalance();

                double tdProfit = newAccountField.getBalance() - preBalance - newAccountField.getDeposit() + newAccountField.getWithdraw();
                if (preBalance == 0) {
                    preBalance = newAccountField.getBalance();
                }
                tdProfitRatioText.setText(String.format("%,.2f%%", tdProfit / preBalance * 100));
                if (tdProfit > 0) {
                    tdProfitRatioText.getStyleClass().add("trade-long-color");
                } else if (tdProfit < 0) {
                    tdProfitRatioText.getStyleClass().add("trade-short-color");
                }
                tdProfitRatioText.setUserData(tdProfit);
            } catch (Exception e) {
                logger.error("渲染错误", e);
            }
            setTodayProfitRatio(tdProfitRatioText);

        }
    }

    private void updateCloseProfit(AccountField newAccountField) {
        if (accountField == null || !CommonUtils.isEquals(accountField.getCloseProfit(), newAccountField.getCloseProfit())) {
            Text closeProfitText = new Text("渲染错误");

            try {
                closeProfitText.setText(String.format("%,.2f", newAccountField.getCloseProfit()));
                if (newAccountField.getCloseProfit() > 0) {
                    closeProfitText.getStyleClass().add("trade-long-color");
                } else if (newAccountField.getCloseProfit() < 0) {
                    closeProfitText.getStyleClass().add("trade-short-color");
                }
                closeProfitText.setUserData(newAccountField);
            } catch (Exception e) {
                logger.error("渲染错误", e);
            }

            setCloseProfit(closeProfitText);
        }
    }

    private void updatePositionProfit(AccountField newAccountField) {
        if (accountField == null || !CommonUtils.isEquals(accountField.getPositionProfit(), newAccountField.getPositionProfit())) {
            Text positionProfitText = new Text("渲染错误");

            try {
                positionProfitText.setText(String.format("%,.2f", newAccountField.getPositionProfit()));
                if (newAccountField.getPositionProfit() > 0) {
                    positionProfitText.getStyleClass().add("trade-long-color");
                } else if (newAccountField.getPositionProfit() < 0) {
                    positionProfitText.getStyleClass().add("trade-short-color");
                }
                positionProfitText.setUserData(newAccountField);
            } catch (Exception e) {
                logger.error("渲染错误", e);
            }
            setPositionProfit(positionProfitText);
        }
    }

    private void updateCommission(AccountField newAccountField) {
        if (accountField == null || !CommonUtils.isEquals(accountField.getCommission(), newAccountField.getCommission())) {
            String commission = "渲染错误";
            try {
                commission = String.format("%,.2f", newAccountField.getCommission());
            } catch (Exception e) {
                logger.error("渲染错误", e);
            }
            setCommission(commission);
        }
    }

    private void updateMargin(AccountField newAccountField) {
        if (accountField == null || !CommonUtils.isEquals(accountField.getMargin(), newAccountField.getMargin())) {
            String margin = "渲染错误";
            try {
                margin = String.format("%,.2f", newAccountField.getMargin());
            } catch (Exception e) {
                logger.error("渲染错误", e);
            }
            setMargin(margin);
        }
    }

    private void updatePreBalance(AccountField newAccountField) {
        if (accountField == null || !CommonUtils.isEquals(accountField.getPreBalance(), newAccountField.getPreBalance())) {
            String preBalance = "渲染错误";
            try {
                preBalance = String.format("%,.2f", newAccountField.getPreBalance());
            } catch (Exception e) {
                logger.error("渲染错误", e);
            }
            setPreBalance(preBalance);
        }
    }

    private void updateDeposit(AccountField newAccountField) {
        if (accountField == null || !CommonUtils.isEquals(accountField.getDeposit(), newAccountField.getDeposit())) {
            String deposit = "渲染错误";
            try {
                deposit = String.format("%,.2f", newAccountField.getDeposit());
            } catch (Exception e) {
                logger.error("渲染错误", e);
            }
            setDeposit(deposit);
        }
    }

    private void updateWithdraw(AccountField newAccountField) {
        if (accountField == null || !CommonUtils.isEquals(accountField.getWithdraw(), newAccountField.getWithdraw())) {
            String withdraw = "渲染错误";
            try {
                withdraw = String.format("%,.2f", newAccountField.getWithdraw());
            } catch (Exception e) {
                logger.error("渲染错误", e);
            }
            setWithdraw(withdraw);
        }
    }

    private void updateGatewayId(AccountField newAccountField) {
        setGatewayId(newAccountField.getGatewayId());
    }

    public boolean isSelected() {
        return selected.get();
    }

    public SimpleBooleanProperty selectedProperty() {
        return selected;
    }

    public void setSelected(boolean selected) {
        this.selected.set(selected);
    }

    public String getAccountId() {
        return accountId.get();
    }

    public void setAccountId(String accountId) {
        this.accountId.set(accountId);
    }

    public SimpleStringProperty accountIdProperty() {
        return accountId;
    }

    public String getCode() {
        return code.get();
    }

    public void setCode(String code) {
        this.code.set(code);
    }

    public SimpleStringProperty codeProperty() {
        return code;
    }

    public String getHolder() {
        return holder.get();
    }

    public void setHolder(String holder) {
        this.holder.set(holder);
    }

    public SimpleStringProperty holderProperty() {
        return holder;
    }

    public String getCurrency() {
        return currency.get();
    }

    public void setCurrency(String currency) {
        this.currency.set(currency);
    }

    public SimpleStringProperty currencyProperty() {
        return currency;
    }

    public String getBalance() {
        return balance.get();
    }

    public void setBalance(String balance) {
        this.balance.set(balance);
    }

    public SimpleStringProperty balanceProperty() {
        return balance;
    }

    public String getMarginRatio() {
        return marginRatio.get();
    }

    public void setMarginRatio(String marginRatio) {
        this.marginRatio.set(marginRatio);
    }

    public SimpleStringProperty marginRatioProperty() {
        return marginRatio;
    }

    public String getAvailable() {
        return available.get();
    }

    public void setAvailable(String available) {
        this.available.set(available);
    }

    public SimpleStringProperty availableProperty() {
        return available;
    }

    public Text getTodayProfit() {
        return todayProfit.get();
    }

    public void setTodayProfit(Text todayProfit) {
        this.todayProfit.set(todayProfit);
    }

    public SimpleObjectProperty<Text> todayProfitProperty() {
        return todayProfit;
    }

    public Text getTodayProfitRatio() {
        return todayProfitRatio.get();
    }

    public void setTodayProfitRatio(Text todayProfitRatio) {
        this.todayProfitRatio.set(todayProfitRatio);
    }

    public SimpleObjectProperty<Text> todayProfitRatioProperty() {
        return todayProfitRatio;
    }

    public Text getCloseProfit() {
        return closeProfit.get();
    }

    public void setCloseProfit(Text closeProfit) {
        this.closeProfit.set(closeProfit);
    }

    public SimpleObjectProperty<Text> closeProfitProperty() {
        return closeProfit;
    }

    public Text getPositionProfit() {
        return positionProfit.get();
    }

    public void setPositionProfit(Text positionProfit) {
        this.positionProfit.set(positionProfit);
    }

    public SimpleObjectProperty<Text> positionProfitProperty() {
        return positionProfit;
    }

    public String getCommission() {
        return commission.get();
    }

    public void setCommission(String commission) {
        this.commission.set(commission);
    }

    public SimpleStringProperty commissionProperty() {
        return commission;
    }

    public String getMargin() {
        return margin.get();
    }

    public void setMargin(String margin) {
        this.margin.set(margin);
    }

    public SimpleStringProperty marginProperty() {
        return margin;
    }

    public String getPreBalance() {
        return preBalance.get();
    }

    public void setPreBalance(String preBalance) {
        this.preBalance.set(preBalance);
    }

    public SimpleStringProperty preBalanceProperty() {
        return preBalance;
    }

    public String getDeposit() {
        return deposit.get();
    }

    public void setDeposit(String deposit) {
        this.deposit.set(deposit);
    }

    public SimpleStringProperty depositProperty() {
        return deposit;
    }

    public String getWithdraw() {
        return withdraw.get();
    }

    public void setWithdraw(String withdraw) {
        this.withdraw.set(withdraw);
    }

    public SimpleStringProperty withdrawProperty() {
        return withdraw;
    }

    public String getGatewayId() {
        return gatewayId.get();
    }

    public void setGatewayId(String gatewayId) {
        this.gatewayId.set(gatewayId);
    }

    public SimpleStringProperty gatewayIdProperty() {
        return gatewayId;
    }
}

File: rt-desktop/src/main/kotlin/xyz/redtorch/desktop/gui/bean/AccountFXBean.kt
package xyz.redtorch.desktop.gui.bean


import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.property.SimpleStringProperty
import javafx.scene.text.Text
import org.slf4j.LoggerFactory
import xyz.redtorch.common.trade.dto.Account
import xyz.redtorch.common.utils.CommonUtils


class AccountFXBean(account: Account, selected: Boolean) {
    companion object {
        private val logger = LoggerFactory.getLogger(AccountFXBean::class.java)
    }

    private val selected = SimpleBooleanProperty(false)
    private val accountId = SimpleStringProperty()
    private val code = SimpleStringProperty()
    private val name = SimpleStringProperty()
    private val currency = SimpleStringProperty()
    private val balance = SimpleStringProperty()
    private val marginRatio = SimpleStringProperty()
    private val available = SimpleStringProperty()
    private val todayProfit = SimpleObjectProperty<Text>()
    private val todayProfitRatio = SimpleObjectProperty<Text>()
    private val closeProfit = SimpleObjectProperty<Text>()
    private val positionProfit = SimpleObjectProperty<Text>()
    private val commission = SimpleStringProperty()
    private val margin = SimpleStringProperty()
    private val preBalance = SimpleStringProperty()
    private val deposit = SimpleStringProperty()
    private val withdraw = SimpleStringProperty()
    private val gatewayId = SimpleStringProperty()

    var accountDTO: Account? = null
        private set

    init {
        update(account, selected)
    }

    fun update(newAccountDTO: Account, newSelected: Boolean) {
        // 更新是否选中
        updateSelected(newSelected)

        if (accountDTO == null) {
            updateAccountId(newAccountDTO)
            updateCode(newAccountDTO)
            updateName(newAccountDTO)
            updateCurrency(newAccountDTO)
            updateGatewayId(newAccountDTO)
        }

        // 对比内存地址
        if (newAccountDTO !== accountDTO) {
            updateChangeable(newAccountDTO)
            accountDTO = newAccountDTO
        }
    }

    private fun updateChangeable(newAccountDTO: Account) {
        updateBalance(newAccountDTO)
        updateMarginRatio(newAccountDTO)
        updateAvailable(newAccountDTO)
        updateTodayProfit(newAccountDTO)
        updateTodayProfitRatio(newAccountDTO)
        updateCloseProfit(newAccountDTO)
        updatePositionProfit(newAccountDTO)
        updateCommission(newAccountDTO)
        updateMargin(newAccountDTO)
        updatePreBalance(newAccountDTO)
        updateDeposit(newAccountDTO)
        updateWithdraw(newAccountDTO)
    }

    private fun updateSelected(selected: Boolean) {
        setSelected(selected)
    }

    private fun updateAccountId(newAccountDTO: Account) {
        setAccountId(newAccountDTO.accountId)
    }

    private fun updateCode(newAccountDTO: Account) {
        setCode(newAccountDTO.code)
    }

    private fun updateName(newAccountDTO: Account) {
        setName(newAccountDTO.name)
    }

    private fun updateCurrency(newAccountDTO: Account) {
        setCurrency(newAccountDTO.currency.toString())
    }

    private fun updateBalance(newAccountDTO: Account) {
        if (accountDTO == null || CommonUtils.isNotEquals(accountDTO!!.balance, newAccountDTO.balance)) {
            setBalance(CommonUtils.formatDouble(newAccountDTO.balance))
        }
    }

    private fun updateMarginRatio(newAccountDTO: Account) {
        if (accountDTO == null //
            || CommonUtils.isNotEquals(accountDTO!!.balance, newAccountDTO.balance) //
            || CommonUtils.isNotEquals(accountDTO!!.margin, newAccountDTO.margin)
        ) {
            setMarginRatio(CommonUtils.formatDouble(newAccountDTO.margin / newAccountDTO.balance * 100, "%,.2f%%"))
        }
    }

    private fun updateAvailable(newAccountDTO: Account) {
        if (accountDTO == null || CommonUtils.isNotEquals(accountDTO!!.available, newAccountDTO.available)) {
            setAvailable(CommonUtils.formatDouble(newAccountDTO.available))
        }
    }

    private fun updateTodayProfit(newAccountDTO: Account) {
        if (accountDTO == null || CommonUtils.isNotEquals(accountDTO!!.balance, newAccountDTO.balance)) {
            val tdProfitText = Text("渲染错误")
            try {
                val preBalance: Double = newAccountDTO.preBalance
                val tdProfit: Double = newAccountDTO.balance - preBalance - newAccountDTO.deposit + newAccountDTO.withdraw
                tdProfitText.text = CommonUtils.formatDouble(tdProfit)
                if (tdProfit > 0) {
                    tdProfitText.styleClass.add("trade-long-color")
                } else if (tdProfit < 0) {
                    tdProfitText.styleClass.add("trade-short-color")
                }
                tdProfitText.userData = tdProfit
            } catch (e: Exception) {
                logger.error("渲染异常", e)
            }
            setTodayProfit(tdProfitText)
        }
    }

    private fun updateTodayProfitRatio(newAccountDTO: Account) {
        if (accountDTO == null || CommonUtils.isNotEquals(accountDTO!!.balance, newAccountDTO.balance)) {
            val tdProfitRatioText = Text("渲染异常")
            try {
                var preBalance: Double = newAccountDTO.preBalance
                val tdProfit: Double = newAccountDTO.balance - preBalance - newAccountDTO.deposit + newAccountDTO.withdraw
                if (preBalance == 0.0) {
                    preBalance = newAccountDTO.balance
                }
                tdProfitRatioText.text = String.format("%,.2f%%", tdProfit / preBalance * 100)
                if (tdProfit > 0) {
                    tdProfitRatioText.styleClass.add("trade-long-color")
                } else if (tdProfit < 0) {
                    tdProfitRatioText.styleClass.add("trade-short-color")
                }
                tdProfitRatioText.userData = tdProfit
            } catch (e: Exception) {
                logger.error("渲染异常", e)
            }
            setTodayProfitRatio(tdProfitRatioText)
        }
    }

    private fun updateCloseProfit(newAccountDTO: Account) {
        if (accountDTO == null || CommonUtils.isNotEquals(accountDTO!!.closeProfit, newAccountDTO.closeProfit)) {
            val closeProfitText = Text(CommonUtils.formatDouble(newAccountDTO.closeProfit))
            if (newAccountDTO.closeProfit > 0) {
                closeProfitText.styleClass.add("trade-long-color")
            } else if (newAccountDTO.closeProfit < 0) {
                closeProfitText.styleClass.add("trade-short-color")
            }
            closeProfitText.userData = newAccountDTO
            setCloseProfit(closeProfitText)
        }
    }

    private fun updatePositionProfit(newAccountDTO: Account) {
        if (accountDTO == null || CommonUtils.isNotEquals(accountDTO!!.positionProfit, newAccountDTO.positionProfit)) {
            val positionProfitText = Text(CommonUtils.formatDouble(newAccountDTO.positionProfit))
            if (newAccountDTO.positionProfit > 0) {
                positionProfitText.styleClass.add("trade-long-color")
            } else if (newAccountDTO.positionProfit < 0) {
                positionProfitText.styleClass.add("trade-short-color")
            }
            positionProfitText.userData = newAccountDTO
            setPositionProfit(positionProfitText)
        }
    }

    private fun updateCommission(newAccountDTO: Account) {
        if (accountDTO == null || CommonUtils.isNotEquals(accountDTO!!.commission, newAccountDTO.commission)) {
            setCommission(CommonUtils.formatDouble(newAccountDTO.commission))
        }
    }

    private fun updateMargin(newAccountDTO: Account) {
        if (accountDTO == null || CommonUtils.isNotEquals(accountDTO!!.margin, newAccountDTO.margin)) {
            setMargin(CommonUtils.formatDouble(newAccountDTO.margin))
        }
    }

    private fun updatePreBalance(newAccountDTO: Account) {
        if (accountDTO == null || CommonUtils.isNotEquals(accountDTO!!.preBalance, newAccountDTO.preBalance)) {
            setPreBalance(CommonUtils.formatDouble(newAccountDTO.preBalance))
        }
    }

    private fun updateDeposit(newAccountDTO: Account) {
        if (accountDTO == null || CommonUtils.isNotEquals(accountDTO!!.deposit, newAccountDTO.deposit)) {
            setDeposit(CommonUtils.formatDouble(newAccountDTO.deposit))
        }
    }

    private fun updateWithdraw(newAccountDTO: Account) {
        if (accountDTO == null || CommonUtils.isNotEquals(accountDTO!!.withdraw, newAccountDTO.withdraw)) {
            setWithdraw(CommonUtils.formatDouble(newAccountDTO.withdraw))
        }
    }

    private fun updateGatewayId(newAccountDTO: Account) {
        setGatewayId(newAccountDTO.gatewayId)
    }

    fun isSelected(): Boolean {
        return selected.get()
    }

    fun selectedProperty(): SimpleBooleanProperty {
        return selected
    }

    fun setSelected(selected: Boolean) {
        this.selected.set(selected)
    }

    fun getAccountId(): String {
        return accountId.get()
    }

    fun setAccountId(accountId: String) {
        this.accountId.set(accountId)
    }

    fun accountIdProperty(): SimpleStringProperty {
        return accountId
    }

    fun getCode(): String {
        return code.get()
    }

    fun setCode(code: String) {
        this.code.set(code)
    }

    fun codeProperty(): SimpleStringProperty {
        return code
    }

    fun getName(): String {
        return name.get()
    }

    fun setName(name: String) {
        this.name.set(name)
    }

    fun nameProperty(): SimpleStringProperty {
        return name
    }

    fun getCurrency(): String {
        return currency.get()
    }

    fun setCurrency(currency: String) {
        this.currency.set(currency)
    }

    fun currencyProperty(): SimpleStringProperty {
        return currency
    }

    fun getBalance(): String {
        return balance.get()
    }

    fun setBalance(balance: String) {
        this.balance.set(balance)
    }

    fun balanceProperty(): SimpleStringProperty {
        return balance
    }

    fun getMarginRatio(): String {
        return marginRatio.get()
    }

    fun setMarginRatio(marginRatio: String) {
        this.marginRatio.set(marginRatio)
    }

    fun marginRatioProperty(): SimpleStringProperty {
        return marginRatio
    }

    fun getAvailable(): String {
        return available.get()
    }

    fun setAvailable(available: String) {
        this.available.set(available)
    }

    fun availableProperty(): SimpleStringProperty {
        return available
    }

    fun getTodayProfit(): Text {
        return todayProfit.get()
    }

    fun setTodayProfit(todayProfit: Text) {
        this.todayProfit.set(todayProfit)
    }

    fun todayProfitProperty(): SimpleObjectProperty<Text> {
        return todayProfit
    }

    fun getTodayProfitRatio(): Text {
        return todayProfitRatio.get()
    }

    fun setTodayProfitRatio(todayProfitRatio: Text) {
        this.todayProfitRatio.set(todayProfitRatio)
    }

    fun todayProfitRatioProperty(): SimpleObjectProperty<Text> {
        return todayProfitRatio
    }

    fun getCloseProfit(): Text {
        return closeProfit.get()
    }

    fun setCloseProfit(closeProfit: Text) {
        this.closeProfit.set(closeProfit)
    }

    fun closeProfitProperty(): SimpleObjectProperty<Text> {
        return closeProfit
    }

    fun getPositionProfit(): Text {
        return positionProfit.get()
    }

    fun setPositionProfit(positionProfit: Text) {
        this.positionProfit.set(positionProfit)
    }

    fun positionProfitProperty(): SimpleObjectProperty<Text> {
        return positionProfit
    }

    fun getCommission(): String {
        return commission.get()
    }

    fun setCommission(commission: String) {
        this.commission.set(commission)
    }

    fun commissionProperty(): SimpleStringProperty {
        return commission
    }

    fun getMargin(): String {
        return margin.get()
    }

    fun setMargin(margin: String) {
        this.margin.set(margin)
    }

    fun marginProperty(): SimpleStringProperty {
        return margin
    }

    fun getPreBalance(): String {
        return preBalance.get()
    }

    fun setPreBalance(preBalance: String) {
        this.preBalance.set(preBalance)
    }

    fun preBalanceProperty(): SimpleStringProperty {
        return preBalance
    }

    fun getDeposit(): String {
        return deposit.get()
    }

    fun setDeposit(deposit: String) {
        this.deposit.set(deposit)
    }

    fun depositProperty(): SimpleStringProperty {
        return deposit
    }

    fun getWithdraw(): String {
        return withdraw.get()
    }

    fun setWithdraw(withdraw: String) {
        this.withdraw.set(withdraw)
    }

    fun withdrawProperty(): SimpleStringProperty {
        return withdraw
    }

    fun getGatewayId(): String {
        return gatewayId.get()
    }

    fun setGatewayId(gatewayId: String) {
        this.gatewayId.set(gatewayId)
    }

    fun gatewayIdProperty(): SimpleStringProperty {
        return gatewayId
    }
}
